chaincode/marbles_transfer/marbles_transfer_queries.go

```diff
@@ -106,16 +111,115 @@ func queryMarblePrice(ctx contractapi.TransactionContextInterface, marbleID stri
 	return string(marblePriceJSON), nil
 }
 
-// TODO add a query to get all of an organization's proposed sales
-// Use GetPrivateDataByPartialCompositeKey to find all keys starting with typeMarbleForSale
-// hint: see sample at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L458
+// QueryMarbleProposedSales returns all of an organization's proposed sales
+func (s *SmartContract) QueryMarbleProposedSales(ctx contractapi.TransactionContextInterface) ([]QueryResult, error) {
+	return queryTxnByType(ctx, typeMarbleForSale)
+}
+
+// QueryMarbleProposedBuys returns all of an organization's proposed buys
+func (s *SmartContract) QueryMarbleProposedBuys(ctx contractapi.TransactionContextInterface) ([]QueryResult, error) {
+	return queryTxnByType(ctx, typeMarbleBid)
+}
+
+func queryTxnByType(ctx contractapi.TransactionContextInterface, txnType string) ([]QueryResult, error) {
+	collection, err := getImplicitCollectionName(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	salesIterator, err := ctx.GetStub().GetPrivateDataByPartialCompositeKey(collection, txnType, []string{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from private data collection: %s", err.Error())
+	}
+	defer salesIterator.Close()
+
+	sales := []QueryResult{}
+
+	var i int
+	for i = 0; salesIterator.HasNext(); i++ {
+		resp, err := salesIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+
+		newMarble := new(Marble)
+		err = json.Unmarshal(resp.Value, newMarble)
+		if err != nil {
+			return nil, err
+		}
+
+		sale := QueryResult{Key: resp.Key, Record: newMarble}
+		sales = append(sales, sale)
+	}
 
-// TODO add a query to get all of an organization's proposed buys
-// Use GetPrivateDataByPartialCompositeKey to find all keys starting with typeMarbleBid
+	return sales, nil
+}
 
 // TODO add a JSON index and query to return all of an organization's marbles larger than a certain size (only works when using CouchDB state database)
 // hint: see sample index at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/META-INF/statedb/couchdb/indexes/indexOwner.json
 // hint: see sample query at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L515
 
-// TODO add a history query so that users can see the chain of custody for a marble since issuance
-// hint: see sample at https://github.com/hyperledger/fabric-samples/blob/master/chaincode/marbles02/go/marbles_chaincode.go#L692
+// QueryMarbleHistory returns the chain of custody for a marble since issuance
+func (s *SmartContract) QueryMarbleHistory(ctx contractapi.TransactionContextInterface, marbleID string) (string, error) {
+	resultsIterator, err := ctx.GetStub().GetHistoryForKey(marbleID)
+	if err != nil {
+		return "", err
+	}
+	defer resultsIterator.Close()
+
+	// buffer is a JSON array containing historic values for the marble
+	var buffer bytes.Buffer
+	buffer.WriteString("[")
```

Comment:
I realize the other sample built JSON manually, but going forward we should use native Go JSON marshal() support.

Reply:
🛠️